File: telecom.R
# load packages
install.packages("readxl")
library(readxl)
install.packages("readr")
library(readr)
# load cluster package for daisy function
install.packages("cluster")
library(cluster)
install.packages("fastDummies")
library(fastDummies)

# load data
telecomData <- read_excel("C:/Dataset/Cluster/Telco_customer_churn.xlsx")

# checking for null data
sum(is.na(telecomData))

# avoid unwanted data cols
updatedTelecomData <- telecomData[,-c(2,3,4,5,6,8,10,11,12,13,14,15,16,17,18,19,20,21,22)]

# create dummy data 
processedTelecomData <- dummy_cols(updatedTelecomData, 
                              select_columns = c("Offer","Paperless Billing","Payment Method"), 
                              remove_first_dummy =  TRUE, 
                              remove_most_frequent_dummy = FALSE, 
                              remove_selected_columns = TRUE)

summary(processedTelecomData)
# distance matrix
distTelecomData <- dist(processedTelecomData, method = "euclidean")

clusterData <- hclust(distTelecomData, method = "complete")

# plot the data - Dendrogram
plot(clusterData, hang = -1)

groupData <- cutree(clusterData, k = 4)

rect.hclust(clusterData, k = 4, border = 'red')

# creating final data
data <- as.matrix(groupData)
finalData <- data.frame(data, updatedTelecomData)

# write to file
write.csv(finalData, "telecom.csv")

aggregate(finalData, list(finalData$data), mean)
# so group 4 gives maximum business while group 1 gives minimum n=business

File: telecom.py
# load packages
import pandas as pd
import scipy.cluster.hierarchy as sch
from sklearn.cluster import AgglomerativeClustering

# load data
telecomData = pd.read_excel(r"C:\Dataset\Cluster\Telco_customer_churn.xlsx")

# checking for na value
telecomData.isna()
telecomData.describe()
telecomData.info()

# avoid unwanted data cols
updatedTelecomData = telecomData.drop(columns = ["Count","Quarter","Referred a Friend","Number of Referrals","Tenure in Months","Phone Service","Multiple Lines","Internet Service","Internet Type","Avg Monthly GB Download","Online Security","Online Backup","Device Protection Plan","Premium Tech Support","Streaming TV","Streaming Movies","Streaming Music","Unlimited Data","Contract"], axis = 1)

# get dummy var data
listOfCol = ["Offer","Paperless Billing","Payment Method"]
dummyTelecomData = pd.get_dummies(data= updatedTelecomData,columns = listOfCol)

dummyTelecomData.describe()

# standardize data
def colVariableStandardization(colVariable):
    return (colVariable - colVariable.mean()/colVariable.std())
stdData = colVariableStandardization(dummyTelecomData.iloc[:, 1:])

# distance matrix - pairwise distance, dendrogram
gData = sch.linkage(stdData, method = 'complete', metric = 'euclidean')

sch.dendrogram(gData)

# agglomerative clustering

data = AgglomerativeClustering(n_clusters= 4, affinity='euclidean', linkage='complete').fit(gData)
data.labels_

clusterLabels = pd.Series(data.labels_)
clusterLabels

telecomData["telecomCluster"] = clusterLabels

cols = telecomData.columns.tolist()
cols
cols = cols[-1:] + cols[:-1]
telecomData = telecomData[cols]
finalData = telecomData

finalData.to_csv("telecomData", encoding = 'utf-8')
